apollo-router-core/src/plugin_utils/structures/router_response.rs

```diff
@@ -0,0 +1,102 @@
+use std::sync::Arc;
+
+use crate::{Context, Error, Object, Path};
+use http::StatusCode;
+use http::{Request, Response};
+use serde_json_bytes::{ByteString, Value};
+
+type CompatRequest = Arc<crate::http_compat::Request<crate::Request>>;
+
+#[derive(Default, Clone)]
+pub struct RouterResponseBuilder {
+    label: Option<String>,
+    data: Option<Value>,
+    path: Option<Path>,
+    has_next: Option<bool>,
+    errors: Vec<Error>,
+    extensions: Option<Object>,
+    context: Option<Context<CompatRequest>>,
+}
+
+impl RouterResponseBuilder {
+    pub fn new() -> Self {
+        Default::default()
+    }
+    pub fn build(&self) -> crate::RouterResponse {
+        self.with_status(StatusCode::OK)
+    }
+    pub fn with_status(&self, status: StatusCode) -> crate::RouterResponse {
+        let this = self.clone();
+        crate::RouterResponse {
+            response: Response::builder()
+                .status(status)
+                .body(
+                    crate::Response {
+                        label: this.label,
+                        data: this.data.unwrap_or_default(),
+                        path: this.path,
+                        has_next: this.has_next,
+                        errors: this.errors,
+                        extensions: this.extensions.unwrap_or_default(),
+                    }
+                    .into(),
+                )
+                .expect("crate::Response implements Serialize; qed")
+                .into(),
+            context: this.context.unwrap_or_else(|| {
+                Context::new().with_request(Arc::new(
+                    Request::new(crate::Request {
```

Comment:
Maybe we could derive `Default` on `Request`

Reply:
We could! let's open a followup issue so we can all agree on that

Reply:
followed up on #437